File: Problem 014 - Longest Collatz sequence.cpp
// Find the number with the longest collatz sequence under 1000000.
#include <iostream>
#include <vector>
#define LIMIT 1000000

using namespace std;

int main(){
    // The current number we're working with
    long long number = 0;
    // The longest chain we've found so far
    int largestChain = 0;
    // The number that caused the longest chain we've found so far
    int largestNumber = 0;
    // Previous chains. Keep this to shorten future searches.
    vector<int> collatz = {0, 1, 2};
    for (int i = 3; i < LIMIT; i++){
        number = i;
        int chain = 0;
        // keep looping until we hit our target.
        while (number != 1){
            // If it is even, divide it by 2.
            if (number % 2 == 0){
                number /= 2;
            }
            // if it is odd, multiply by 3 then add 1.
            else {
                number = ((number * 3) + 1);
            }
            // Lengthen the chain by 1.
            chain++;
            // If we've hit a number we've had before we should already know how long that number's chain was. Add that to our current chain, then stop.
            if (number < i){
                chain += collatz.at(number);
                if (chain > largestChain){
                    largestChain = chain;
                    largestNumber = i;
                }
                // And remember how long the chain for this number was to shorten future computations.
                collatz.push_back(chain);
                break;
            }
        }
    }
    cout<<"The number under "<<LIMIT<<" with the largest chain is: "<<largestNumber<<" at "<<largestChain<<" long.";
    return 0;
}

File: LUA/Problem014 - Longest Collatz sequence.lua
--[[
The following iterative sequence is defined for the set of positive integers:

n ? n/2 (n is even)
n ? 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:

13 ? 40 ? 20 ? 10 ? 5 ? 16 ? 8 ? 4 ? 2 ? 1
It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.
]]

-- Variables: Previous chain lengths; upper limit for our search; largest chain found so far; number that produced the longest chain so far
collatz = {1, 2}
limit = 1000000
largestChain = 0
largestNumber = 0

-- Look at every number below our limit (except 1 and 2 since we know those have chains of length 1 and 2 respectively)
for i = 3, limit - 1 do
	local chain = 0
	local number = i
	-- Until we hit 1 with our number (or if we later find a number we've seen before) we keep going
	while number ~= 1 do
		-- Collatz sequence: divide by 2 if even, multiply by 3 and add 1 if odd.
		if number % 2 == 0 then
			number = number / 2
		else
			number = (number * 3) + 1
		end
		chain = chain + 1
		-- If we've seen this number before, we should know what the sequence length is for it, so add that length and take an early out
		if number < i then
			chain = chain + collatz[number]
			break
		end
	end
	-- If this chain is the largest so far, remember it and the number that caused it
	if chain > largestChain then
		largestChain = chain
		largestNumber = i
	end
	-- Insert the chain for this number into our table for future use
	table.insert(collatz, chain)
end

-- Write our final answer
io.write(string.format("The number below %d that produced the largest collatz sequence was %d, with a sequence of length %d", limit, largestNumber, largestChain))
